### PR TITLE
bpo-46263: FreeBSD 14.0 jemalloc workaround for junk bytes of freed memory

### PR DESCRIPTION
**Pull request by tiran**
https://bugs.python.org/issue46263

Automerge-Triggered-By: GH:tiran

### CODE REVIEW

**Review by pablogsal — APPROVED** [review]

**Comment by tiran** [comment]
https://www.freebsd.org/cgi/man.cgi?query=jemalloc&apropos=0&sektion=0&manpath=FreeBSD+14.0-current&arch=default&format=html

>       Additionally, --enable-debug is enabled in development versions of FreeBSD

```
   opt.junk	(const char *) r- [--enable-fill]
	   Junk	filling. If set	to "alloc", each byte of uninitialized
	   allocated memory will be initialized	to 0xa5. If set	to "free", all
	   deallocated memory will be initialized to 0x5a. If set to "true",
	   both	allocated and deallocated memory will be initialized, and if
	   set to "false", junk	filling	be disabled entirely. This is intended
	   for debugging and will impact performance negatively. This option
	   is "false" by default unless	--enable-debug is specified during
	   configuration, in which case	it is "true" by	default.

```

File: Lib/test/test_capi.py [hunk 1]
```diff
@@ -857,8 +857,13 @@ class PyMemDebugTests(unittest.TestCase):
 
     def check(self, code):
         with support.SuppressCrashReport():
-            out = assert_python_failure('-c', code,
-                                        PYTHONMALLOC=self.PYTHONMALLOC)
+            out = assert_python_failure(
+                '-c', code,
+                PYTHONMALLOC=self.PYTHONMALLOC,
+                # FreeBSD: instruct jemalloc to not fill freed() memory
```
Comment (gpshead) [inline Lib/test/test_capi.py:863]: this is a rather fragile test as written. depending on the undefined behavior of what the contents of free()d memory will be is our problem.  jemalloc is doing something perfectly reasonable, it's our test that is not. expect this to come up again at a random future point even if this workaround does something useful today.

Comment (tiran) [reply Lib/test/test_capi.py:863]: Other memory allocators may use different constants. I have an open PR for mimalloc https://github.com/microsoft/mimalloc/pull/517

Comment (vstinner) [reply Lib/test/test_capi.py:863]: Maybe _PyMem_IsPtrFreed() constants could be updated at runtime if Python detects when jemalloc is running?

I already had to change _PyMem_IsPtrFreed() constants when I implemented this function, to use Microsoft memory allocator constants.

_PyMem_IsPtrFreed() is only called by _PyObject_IsFreed() which is only used when Python is built in debug mode (and a few _testcapi test functions in release mode as well). So performance is not critical here.

Comment (tiran) [reply Lib/test/test_capi.py:863]: You would have to detect that you run under jemalloc **and** jemalloc is built with debug flags **or** jemalloc is configured to taint freed memory. FreeBSD 14.0 is under development and its libc uses a debug build. Release builds of FreeBSD are not affected.

Comment (gpshead) [reply Lib/test/test_capi.py:863]: A lot of memory allocators do this when not building release/optimized builds. It isn't just jemalloc. I don't see an obvious clean solution.

**Pull request merged**